File: server/haskell/src/Record/Get.hs
module Record.Get (getRecord) where

import Auth.JWT (authorizedUserId)
import DB (execute)
import Data.Aeson.TH (defaultOptions, deriveJSON)
import Data.Int (Int64)
import qualified Data.IntMap.Strict as IntMap
import Data.Set (Set)
import qualified Data.Set as Set
import qualified Data.Text as T
import Data.Time (UTCTime)
import qualified Data.Vector as V
import qualified Hasql.Pool as HP
import qualified Hasql.Statement as S
import qualified Hasql.TH as TH
import Network.HTTP.Types (status200, status500)
import Record.Go (Color (Black, White), identifyCaptures, runBoardA, toCoord')
import Web.Scotty
  ( ScottyM,
    get,
    json,
    param,
    raiseStatus,
    status,
  )

data Point = Point {x :: Int, y :: Int} deriving (Show)

$(deriveJSON defaultOptions ''Point)

data Move = Move {position :: Maybe Point, color :: String, captures :: [Point]} deriving (Show)

$(deriveJSON defaultOptions ''Move)

data Stone = Stone {x :: Int, y :: Int, color :: String} deriving (Show)

$(deriveJSON defaultOptions ''Stone)

data GetResponse = GetResponse
  { id :: Int,
    owner :: Int,
    board_size :: Int,
    created :: UTCTime,
    name :: String,
    black_player :: String,
    white_player :: String,
    comment :: String,
    handicap :: Int,
    komi :: Double,
    ruleset :: String,
    winner :: String,
    -- TODO moves
    moves :: [Move],
    stones :: [Stone]
  }
  deriving (Show)

$(deriveJSON defaultOptions ''GetResponse)

type RecordRow = (Int64, T.Text, T.Text, T.Text, T.Text, Int64, Double, T.Text, T.Text, UTCTime)

selectRecord :: S.Statement (Int64, Int64) RecordRow
selectRecord =
  [TH.singletonStatement|
    select
    board_size :: int8, name :: text, black_player :: text, white_player :: text, comment :: text, handicap :: int8, komi :: float8, ruleset :: text, winner :: text, created :: timestamptz
    from record_record
    where
    owner_id = $1 :: int8 and id = $2 :: int8
  |]

type MoveRow = (Maybe Int64, T.Text)

selectMoves :: S.Statement Int64 (V.Vector MoveRow)
selectMoves =
  [TH.vectorStatement|
    select
    position :: int8?, color :: text
    from record_move
    where
    record_id = $1 :: int8
  |]

fromColor :: Color -> String
fromColor color = case color of
  Black -> "B"
  White -> "W"

toResponse :: Int64 -> Int64 -> RecordRow -> [((Maybe Int, Color), Set Int)] -> [((Int, Int), Color)] -> GetResponse
toResponse userId recordId (board_size, name, black_player, white_player, comment, handicap, komi, ruleset, winner, created) moves stones =
  GetResponse
    { id = fromIntegral recordId,
      owner = fromIntegral userId,
      board_size = fromIntegral board_size,
      created,
      name = T.unpack name,
      black_player = T.unpack black_player,
      white_player = T.unpack white_player,
      comment = T.unpack comment,
      handicap = fromIntegral handicap,
      komi,
      ruleset = T.unpack ruleset,
      winner = T.unpack winner,
      moves =
        [ Move
            { position = fmap (uncurry Point . toCoord' (fromIntegral board_size)) position,
              color = fromColor color,
              captures = [Point {x, y} | (x, y) <- map (toCoord' $ fromIntegral board_size) $ Set.toAscList captures]
            }
          | ((position, color), captures) <- moves
        ],
      stones =
        [ Stone
            { x,
              y,
              color = if color == Black then "B" else "W"
            }
          | ((x, y), color) <- stones
        ]
    }

getRecord :: HP.Pool -> ScottyM ()
getRecord pool = get "/api/records/:recordId/" $ do
  userId <- authorizedUserId
  recordId <- param "recordId"
  record <- execute pool selectRecord (userId, recordId)
  moves' <- execute pool selectMoves recordId
  let moves = V.toList moves'
  -- TODO desgostang check if you can extract non-tuples from hasql
  let (size', _, _, _, _, _, _, _, _, _) = record
  let size = fromIntegral size'
  -- TODO effective error checking for invalid move strings
  let movesToPlay = [(fmap fromIntegral position, if color == "B" then Black else White) | (position, color) <- moves]
  case runBoardA size $ identifyCaptures movesToPlay of
    Left _ -> raiseStatus status500 "Invalid game record."
    Right (movesAndCaptures, board) -> do
      let stones = [(toCoord' size pos, color) | (pos, color) <- IntMap.toAscList board]
      status status200
      json $ toResponse userId recordId record movesAndCaptures stones

File: server/rust/src/record/get.rs
use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::IntoResponse,
    Json,
};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use tokio_postgres::Client;

use crate::{
    auth::UserId,
    db::{query, query_one},
    record::go::{play_out_game, Color, Move, Pos},
};

#[derive(Serialize, Deserialize, PartialEq, Eq)]
struct Point {
    x: i32,
    y: i32,
}

impl PartialOrd for Point {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}
impl Ord for Point {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.partial_cmp(other).unwrap()
    }
}

#[derive(Serialize, Deserialize)]
struct MoveResponse {
    position: Option<Point>,
    color: String,
    captures: Vec<Point>,
}

#[derive(Serialize, Deserialize, PartialEq, Eq)]
struct Stone {
    x: i32,
    y: i32,
    color: String,
}
impl PartialOrd for Stone {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}
impl Ord for Stone {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.partial_cmp(other).unwrap()
    }
}

#[derive(Serialize, Deserialize)]
struct GetResponse {
    id: i64,
    owner: i32,
    board_size: i32,
    created: DateTime<Utc>,
    name: String,
    black_player: String,
    white_player: String,
    comment: String,
    handicap: i32,
    komi: f64,
    ruleset: String,
    winner: String,
    moves: Vec<MoveResponse>,
    stones: Vec<Stone>,
}

pub async fn get(
    State(client): State<Arc<Client>>,
    UserId(user_id): UserId,
    Path(record_id): Path<i64>,
) -> Result<impl IntoResponse, (StatusCode, &'static str)> {
    let row = query_one(&client, "SELECT id, board_size, name, black_player, white_player, comment, handicap, komi, ruleset, winner, created FROM record_record WHERE owner_id = $1 AND id = $2", &[&user_id, &record_id]).await?;
    let board_size = row.get("board_size");
    let created = row.get("created");
    let name = row.get("name");
    let black_player = row.get("black_player");
    let white_player = row.get("white_player");
    let comment = row.get("comment");
    let handicap = row.get("handicap");
    let komi = row.get("komi");
    let ruleset = row.get("ruleset");
    let winner = row.get("winner");
    let rows = query(
        &client,
        "SELECT position, color FROM record_move WHERE record_id=$1 ORDER BY move ASC",
        &[&record_id],
    )
    .await?;
    let moves: Vec<Move> = rows
        .iter()
        .map(|row| (row.get::<&str, Option<i32>>("position"), row.get("color")))
        .map(|(pos, color)| (pos.map(Pos::new), Color::new(color)))
        .collect();
    let (board, moves) = play_out_game(board_size, &moves).map_err(|_| {
        (
            StatusCode::INTERNAL_SERVER_ERROR,
            "Error playing moves in the DB",
        )
    })?;
    let moves = moves
        .iter()
        .map(|((pos, color), captures)| {
            let mut captures: Vec<Point> = captures
                .iter()
                .map(|p| p.to_coord(board_size))
                .map(|(x, y)| Point { x, y })
                .collect();
            captures.sort();
            MoveResponse {
                position: pos
                    .as_ref()
                    .map(|p| p.to_coord(board_size))
                    .map(|(x, y)| Point { x, y }),
                color: format!("{}", color),
                captures,
            }
        })
        .collect();
    let mut stones: Vec<Stone> = board
        .iter()
        .map(|(pos, color)| (pos.to_coord(board_size), format!("{}", color)))
        .map(|((x, y), color)| Stone { x, y, color })
        .collect();
    stones.sort();
    Ok(Json(GetResponse {
        id: record_id,
        owner: user_id,
        board_size,
        created,
        name,
        black_player,
        white_player,
        comment,
        handicap,
        komi,
        ruleset,
        winner,
        moves,
        stones,
    }))
}
